pubsub.go

```diff
@@ -710,10 +809,72 @@ func (p *PubSub) Publish(topic string, data []byte) error {
 			return err
 		}
 	}
+
+	p.bootstrap(topic)
 	p.publish <- &Message{m}
+
 	return nil
 }
 
+// bootstrap handles initial bootstrapping of peers for a given topic
+func (p *PubSub) bootstrap(topic string) {
+	_, bootstrapped := p.bootstrapped.LoadOrStore(topic, struct{}{})
+	if !bootstrapped {
```

Comment:
Any reason to call this up-front instead of just letting the router tell us when to call it? This will:

1. Force a discovery even if we have peers.
2. Not discover even if we've disconnected from all relevant peers.

If we just want to wait to connect to peers before publishing, we should consider doing that in the router. That is, have the router synchronously ask for more peers when publishing to an empty peer list.

Reply:
> We may actually discover even if we've disconnected from all peers (unlike the current implementation) since we have a "rediscovery" mechanism that the routers can trigger.

IMO, this should be the _only_ mechanism. That is, routers should have to say "please bootstrap me".

> I'd like to put the logic in PubSub instead of the routers if we can since then we only need to write the code once for all of the routers.

Does pubsub know enough to do this?

Reply:
> Does pubsub know enough to do this?

Yes/It depends. PubSub knows how many peers are subscribed to a given topic and has access to a discovery mechanism. If your router had a data structure that tracked network topology then the way you would signal "I need 2 good peers or 4 bad ones" is by asking for 2 and then asking for more if what you have is insufficient.

It's also worth noting that we may want some of this to come from the application layer. For example, with a persistent pubsub we may feel comfortable publishing to fewer peers than without persistence. This could be because in a persistent pubsub even one peer can eventually get the message out, while in an ephemeral pubsub once a message is processed it is dead.

> IMO, this should be the only mechanism. That is, routers should have to say "please bootstrap me".

Not an unreasonable request. It's a bit tricky though since we can't block in the routers as they all share the same event loop. As a result, we may have to do some gross async/callback things. If you think this is the way to go (now as opposed to dealing with this PR first and adding an issue) let me know and I'll dig more into it.